src/cardsort/analysis.py

```diff
@@ -13,19 +14,47 @@
 ]
 
 
-def _get_distance_matrix_for_user(df_user: pd.DataFrame) -> np.ndarray:
-    n = len(df_user)
+logger = logging.getLogger(__name__)
 
-    X = np.zeros((n, n))
-    for i in range(n):
-        for j in range(i, n):
-            cat1 = df_user.loc[df_user["card_id"] == (i + 1), "category_label"].values[
-                0
-            ]
-            cat2 = df_user.loc[df_user["card_id"] == (j + 1), "category_label"].values[
-                0
-            ]
-            X[i, j] = X[j, i] = 0 if cat1 == cat2 else 1
+
+def _check_data(df: pd.DataFrame) -> bool:
+    # check if user_ids go from 1 to n=number of users
+    user_ids = df["user_id"].unique()
+    for i, user_id in enumerate(user_ids):
+        if user_id != i + 1:
+            logger.error("User_ids are not in the right order or contain gaps.")
+            return False
+    # check if each card_id is always associated with exactly one card_label
+    card_id_counts = df.groupby("card_id")["card_label"].nunique()
+    for card_id, count in card_id_counts.items():
+        if count != 1:
+            logger.error(
+                f"Card_id {card_id} is associated with {count} different card_labels."
+            )
+            return False
+    # check if card_ids go from 1 to n=number of cards
+    card_ids = df["card_id"].unique()
+    for i, card_id in enumerate(card_ids):
+        if card_id != i + 1:
+            logger.error("Card_ids are not in the right order or contain gaps.")
+            return False
+    # check if all users categorize each card exactly once
+    categorized_cards_per_user = (
```

Comment:
These lines only checks than there no user who categorised twice a same card_id, it does not check that every user_id categorised every card_id.

You can draw inspiration from this code
```python
    counts = df.groupby(['user_id', 'card_id']).size()
    if (counts > 1).any():
        logger.error("At least one user categorized at least one card more than once.")
    n_cards = df["card_id"].nunique()
    n_users = df["user_id"].nunique()
    expected_size = n_cards * n_users
    if len(counts) != expected_size:
        logger.error("At least one user does not categorized at least one card.")
```

Reply:
Done: a3c958f